EIPS/eip-5568.md

```diff
@@ -0,0 +1,63 @@
+---
+eip: 5568
+title: Revert Signals
+description: Signal to wallets that an action is needed by returning a custom revert code
+author: Pandapip1 (@Pandapip1)
+discussions-to: https://ethereum-magicians.org/t/eip-5568-revert-signals/10622
+status: Draft
+type: Standards Track
+category: Interface
+created: 2022-08-31
+requires: 140
+---
+
+## Abstract
+
+This EIP introduces a machine-readable format for reverts to signal to wallets that an action needs to be taken.
+
+## Motivation
+
+Oftentimes, a smart contract needs to signal to a wallet that an action needs to be taken, such as to sign a transaction or send an HTTP request to a URL. Traditionally, this has been done by hard-coding the logic into the frontend, but this EIP allows the smart contract itself to request the action.
+
+This means that, for example, an exchange or a market can directly tell the wallet to approve the smart contract to spend the token, vastly simplifying the front-end code.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Custom Revert Reason
+
+To send a signal to a wallet, a smart contract MUST revert with the following [EIP-140](./eip-140.md) revert data:
+  
+```solidity
+abi.encodePacked(uint16(EIP_NUMBER), uint24(INSTRUCTION_ID1), uint16(bytes(INSTRUCTION_DATA1).length), bytes(INSTRUCTION_DATA1), uint24(INSTRUCTION_ID2), uint16(bytes(INSTRUCTION_DATA2).length), bytes(INSTRUCTION_DATA2), ...)
+```
+
+`EIP_NUMBER` MUST be `5568`.
+
+The `INSTRUCTION_ID` of an instruction defined by an EIP MUST be its EIP number. An EIP MUST define exactly zero or one `INSTRUCTION_ID`. The structure of the instruction data for any `INSTRUCTION_ID` MUST be defined by the EIP that defines the `INSTRUCTION_ID`.
+
+### Signal Response
+
+Before submitting a transaction to the mempool, it MUST be evaluated in a local fork of the chain. If it reverts and has a revert reason starting with the bytes `0x15C0`, then it MUST be treated as a signal.
+
+To parse the signal, discard the first 2 bytes (the `0x15C0` magic number). Then, repeat the following until the signal is fully read:
+
+Read the next 3 bytes to determine the `INSTRUCTION_ID`, and the next 2 bytes to determine the number of bytes to read. Then, read that many bytes to get the `INSTRUCTION_DATA`. Finally, evaluate the instruction as defined in its specification.
+
+After all the instructions have been followed, the wallet SHOULD NOT re-evaluate the transaction, except if an instruction explicitly states that the transaction should be re-evaluated.
+
+## Rationale
+
+This EIP was explicitly optimized for deployment gas cost. It is expected that libraries will eventually be developed that makes sending and receiving signals more developer-friendly.
+
+## Backwards Compatibility
+
+Needs discussion.
+
+## Security Considerations
+
+Needs discussion.
```

Comment:
IMO that would be more important for future extensions. I don't think this EIP should try to anticipate what other EIPs will or will not do.

Reply:
That's very true. You could at least mention that it might be possible to collide with non-signal revert messages? Or explain how it isn't possible. Whichever :P